config/v1alpha2/types_insights.go

```diff
@@ -0,0 +1,207 @@
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// spec holds user settable values for configuration
+	// +required
+	Spec InsightsDataGatherSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status InsightsDataGatherStatus `json:"status"`
+}
+
+type InsightsDataGatherSpec struct {
+	// gatherConfig is an optional spec attribute that includes all the configuration options related to gathering of the Insights data and its uploading to the ingress.
+	// +optional
+	GatherConfig GatherConfig `json:"gatherConfig"`
+}
+
+type InsightsDataGatherStatus struct{}
+
+// gatherConfig provides data gathering configuration options.
+// +kubebuilder:validation:XValidation:rule="has(self.mode) && self.mode == 'Custom' ?  has(self.custom) : !has(self.custom)",message="custom is required when mode is Custom, and forbidden otherwise"
+type GatherConfig struct {
+	// dataPolicy is an optional list of DataPolicyOptions that allows user to enable additional obfuscation of the Insights archive data.
+	// It may not exceed 2 items and must not contain duplicates.
+	// Valid values are ObfuscateNetworking and WorkloadNames.
+	// When set to ObfuscateNetworking the IP addresses and the cluster domain name are obfuscated.
+	// When set to WorkloadNames the data from Deployment Validation Operator is obfuscated.
+	// When omitted no obfuscation is applied.
+	// +kubebuilder:validation:MaxItems=2
+	// +kubebuilder:validation:XValidation:rule="self.all(x, self.exists_one(y, x == y))",message="DataPolicy must not contain duplicates"
+	// +optional
+	DataPolicy []DataPolicyOption `json:"dataPolicy,omitempty"`
+	// mode is a required field that specifies the mode for gatherers.
+	// When set to Enabled, all gatherers are enabled.
+	// When set to Disabled, all gatherers are disabled.
+	// When set to Custom, the custom configuration from the gatherers field is used.
+	// +required
+	Mode GatheringMode `json:"mode"`
+	// custom is an optional list of gatherers configurations that can be used to enable or disable specific gatherers.
+	// It may not exceed 100 items and each gatherer can be present only once.
+	// When omitted, this means that all gatherers are enabled.
+	// It is possible to disable an entire set of gatherers while allowing a specific function within that set.
+	// For example, you can disable "gathererOne" while enabling "gathererOne/functionOne."
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// +kubebuilder:validation:MaxItems=100
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	Custom []GathererConfig `json:"custom,omitempty"`
```

Comment:
It might be worth making it explicit in the GoDoc for this field that the configuration that is specified is what is used and no defaulting behavior takes place.

Reply:
I have reworked it with a custom struct, so you may check if the documentation is good enough. There should be no defaults anymore